Added OpenWeatherMap service

dickwolff opened the pull request:
## Description

As suggested per issue #180 I created a service to show weather information for a specified location. I used OpenWeatherMap as source since they offer a free API to request forecast. Example below.

![image](https://user-images.githubusercontent.com/5620002/104499910-07fc2d80-55de-11eb-8369-eb3cf6f6c5e9.png)

This service has the following options that are to be provided in the config.yml file:
- location: The city to request the weather forecast from.
- locationId: The specific OpenWeatherMap City ID to request the weather forecast from. Retrieve this from the OWM website by searching for your city and retrieve the ID from the URL. `locationId` is preferred over `location`.
  ![image](https://user-images.githubusercontent.com/5620002/107193012-a4073200-69ee-11eb-9661-977d6b337253.png)

- apiKey: Your private API key to access the OpenWeatherMap API (request [here](https://openweathermap.org/appid)).
- type: Should be "OpenWeather".
- units: (Optional. The units to display temperature. Accepts metric, imperial or kelvin. Defaults to kelvin (which is the OWM default).
- background: Choose which type of background you want behind the image. Can be one of: square, cicle, none. Defaults to none. Added this option because the image could not be distinguished in dark mode.

See example configuration below.
```
- name: "Weather" 
  location: "Amsterdam"
  locationId: "2759794" # Is preferred over `location`
  apiKey: "abc<------------------->xyz"
  units: "metric"
  background: "square"
  type: "OpenWeather"
```

Fixes #180

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] I've read & comply with the [contributing guidelines](https://github.com/bastienwirtz/homer/blob/main/CONTRIBUTING.md)
- [x] I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers. 
- [x] I have made corresponding changes the documentation (README.md).
- [x] I've checked my modifications for any breaking changes, especially in the `config.yml` file

mikeanth-dev commented:
I can't seem to be able to implement this. Should this item go under `services`? This is what i tried:
``` yml
services:
  - name: Info
    icon: fas fa-info
    items:
      - name: Weather
        location: Bucharest
        apiKey: my_api_key_here
        units: metric
        background: square
        type: OpenWeather
```
But i get an empty section within homer when saving and running this.

dickwolff commented:
Your config is correct, see mine below:
```
services:
  - name: "Applications"
    icon: "fas fa-box-open"
    items:
       - name: "Weather"
         url: "https://api.openweathermap.org/data/2.5/weather"
         location: "Amsterdam"
         apiKey: "<---------------------->"
         units: "metric"
         background: "circle"
         type: "OpenWeather"
```

Could you check the console to see if OpenWeatherMaps gives you an error? It could be your location could not be found, or maybe your API key is not yet activated. It took about an hour for my key to be activated.

mikeanth-dev commented:
I don't get any errors in the console. I was, however, missing the `url` part. It is not mentioned in the PR description. However, i still get a blank section. I'll wait on it, maybe my api key is not yet activated. Thank you!

dickwolff commented:
I'm sorry, the URL should have been there. This was from a previous version of the service where the URL was configurable but I decided to make it fixed because there was really no reason why it should be configurable.

If you still have a blank tile, please let me know. I'll try it out myself later today if I can get to it.

mikeanth-dev commented:
Yes, it's still am empty section. If it matters, i am using the latest docker image from dicker hub, and i am running on an arm64 platform.

dickwolff commented:
This feature is not yet merged into the master branch. So if you're using that, it's basically not there yet. You'll have to wait until @bastienwirtz has validated and approved my change 😉

mikeanth-dev commented:
Oh... I see. Is there any ETA on that? @bastienwirtz

dickwolff commented:
I found that not all cities could be found, specifically cities in the US that exists in multiple states. Adding a state to location didn't do anything. So I added the option to add a location by city ID. I've updated my code, and also updated the documentation in `customservices.md` and in this PR.

bastienwirtz commented:
:wave:

@dickwolff @mikeanth-dev Sorry about that, a little project (:baby:) is kind of eating some of my time, I will review & test your PRs soon!
Thanks again for the contribs!

bastienwirtz commented on docs/configuration.md:
````diff
@@ -118,6 +118,8 @@ services:
         # background: red # optional color for card to set color directly without custom stylesheet
 ```
 
+View [Custom Services](customservices.md) for details about all available custom services (like PiHole) and how to configure them.
````
Geat idea! :+1:

bastienwirtz commented on src/components/services/OpenWeather.vue:
```diff
@@ -0,0 +1,164 @@
+      if (this.api) {
+        return this.api.name;
+      }
+      return "";
```
why adding an additional properties when you can access it directly in the template from the component's data (with a default "" value already set)?

DerpgonCz replied:
I would go with `this.api?.name || ''`

bastienwirtz commented on src/components/services/OpenWeather.vue:
```diff
@@ -0,0 +1,164 @@
+  },
+  data: () => ({
+    error: false,
+    api: {
```
`api` is kinda confusing to me, it designate the service, `weaterData` or something like that could be great, or event better, using simple vars instead of an object representing the api response structure ? you could assign those vars in the fetch function, and filter the api result to get only what you need (for exemple, `humidity` is unused here).

bastienwirtz commented on docs/customservices.md:
```diff
@@ -0,0 +1,38 @@
+items:
+     - name: "Weather"
+       location: "Amsterdam" # your location.
+       locationId: "2759794" # OpenWeatherMap city ID.
```
What to you think of
```suggestion
       locationId: "2759794" # Optional: Specify OpenWeatherMap city ID for better accuracy
```

Or something like that. To make it clear that location and locationID are 2 different way to define the same thing.

bastienwirtz commented on src/components/services/OpenWeather.vue:
```diff
@@ -0,0 +1,164 @@
+                    {{ temp }}
+                    <span v-if="this.item.units === 'metric'">&deg;C</span>
+                    <span v-else-if="this.item.units === 'imperial'">&deg;F</span>
+                    <span v-else>K</span>
```
`unitLabel` could be a computed property to avoid the logic on the template level.

DerpgonCz replied:
I would use a filter as follows: `temp | formatTemp`

bastienwirtz reviewed (changes requested):
Great work! I think it would be great to simplify this a bit: 
- Remove all the computed properties (always great to give less work to the dynamic data binding for better performance).
- Declare simple var in `data` for thing you want to display
- Filter api output & assign data vars directly in the `fetchData` function.

How does that sound to you @dickwolff?
Thanks again!

dickwolff commented:
Thanks for your feedback @bastienwirtz! This is the first time I really did something with Vue.js, so I just took inspiration from the PiHole service and tried to make it work 😅. 

I’ll try to make the improvements later this week!

bastienwirtz commented:
Thanks @dickwolff! Let me know if you have any question!

nodecentral commented:
Hi, please could someone confirm the yaml entry to use for this, does it need the fill url etc.

dickwolff commented:
I haven't been able to push the refactored code. I hope to push it by this weekend.

bastienwirtz commented:
@dickwolff no pressure, if you want I can merge it like that and do the adjustment myself if you prefer, let me know! I would be happy to help!

dickwolff commented:
Seems I can't find the time to fix it myself. @bastienwirtz Will you please merge it and improve it yourself? I look forward to the resulting code, hope I learn something from it 🙂

bastienwirtz commented:
Of course @dickwolff I'll look into it asap! Thanks again for your contribution!

dickwolff commented:
Did you manage to improve on my code @bastienwirtz ?
